evm/src/cpu/kernel/asm/ripemd/memory.asm

```diff
@@ -0,0 +1,169 @@
+global ripemd_storage: // starts by initializing buffer
+    // stack: i [init: 64]
+    %store_zeros(64, ripemd_storage)
+    // stack:
+    %jump(store_size)
+
+store_size:
+    // stack: length
+    %shl_const(3)
+    // stack: abcdefgh
+    %extract_and_store_byte(64)
+    // stack: abcdefg
+    %extract_and_store_byte(65)
+    // stack: abcdef
+    %extract_and_store_byte(66)
+    // stack: abcde
+    %extract_and_store_byte(67)
+    // stack: abcd
+    %extract_and_store_byte(68)
+    // stack: abc
+    %extract_and_store_byte(69)
+    // stack: ab
+    %extract_and_store_byte(70)
+    // stack: a
+    %mstore_ripemd_offset(71)
+    // stack:           0x80     // padding has 0x80 in first position and zeros elsewhere
+    %mstore_ripemd_offset(72)    // store first padding term here so as to avoid extra label
+    %jump(store_padding)
+
+store_padding:
+    // stack: i (init 63)
+    %store_zeros(136, store_padding)
+    DUP1
+    %jumpi(store_input_alt)
+    POP
+    %jump(ripemd_init)
+
+store_input_alt:
+    // stack:               rem, length, REM_INP
+    %stack (rem, length, head) -> (length, rem, 136, head, rem, length)
+    SUB
+    ADD
+    // stack: offset, byte, rem, length, REM_INP
+    %mstore_ripemd
+    // stack:               rem, length, REM_INP
+    %sub_const(1)
+    DUP1
+    // stack:  rem - 1, rem - 1, length, REM_INP
+    %jumpi(store_input_alt)
+    // stack:                 0, length
+    POP
+    %jump(ripemd_init)
+
+
+store_input:
+    // stack:               ADDR    , rem    , length
+    DUP3
+    DUP3
+    DUP3
+    MLOAD_GENERAL
+    // stack:         byte, ADDR    , rem    , length 
+    DUP5
+    DUP7
+    SUB
+    %add_const(136)
+    // stack: offset, byte, ADDR    , rem    , length 
+    %mstore_ripemd
+    // stack:               ADDR    , rem    , length 
+    SWAP2
+    %add_const(1)
+    SWAP2
+    // stack:               ADDR + 1, rem    , length
+    SWAP3
+    %sub_const(1)
+    SWAP3
+    // stack:               ADDR + 1, rem - 1, length 
+    DUP4
+    %jumpi(store_input)
+    // stack:               ADDR    , 0      , length
+    %pop4
+    // stack:                                  length
+    %jump(ripemd_init)
+
```

Comment:
nit: spacing (in various places)

Reply:
hmm, I guess this is a purposeful style choice---I wanted to right-align the stack and visually align persistent values on the stack at different steps. I found this helped me keep track of what was happening. Has the team established explicit style standards on such things? @dlubarov

Reply:
Ohh I meant the line spacing — having extra lines (or at least inconsistently) between the functions

Reply:
The stack diagram spacing is fine! It's different form how I do it but neither is correct

Reply:
oh sorry for the misunderstanding! is the rule that there should always be precisely 1 newline between labels/macros?

Reply:
Hm not necessarily, just that it should be consistent! I think typically it would be one newline, unless multiple labels are meant to be grouped together (in which case there would be more newlines between the groups)